enhancements/cluster-scope-secret-volumes/shared-resource-validation.md

```diff
@@ -0,0 +1,413 @@
+discuss as needed.
+
+The OCP teams possibly involved with issues:
+- API Server (webhook invocation, confusion around creating/modifying objects)
+- Kubelet/node (incorrectly perhaps if valid Pods fail to come up, unless there is a problem beyond this EP's validations)
+- Cluster Storage Operator (currently storage team, if a CSO problem leads to the webhook not coming up) 
+- Shared Resource Operator/Driver themselves (currently build/jenkins team)
+
+
+#### Support Procedures
+
+##### Detect Failure modes and Possible Symptoms
+
+- If the webhook is not running, kube-apiserver logs will show errors like "failed to call admission webhook xyz".
+- Cluster Storage Operator / CSI Shared Resource Operator will degrade with message "Failed to launch webhook server" and reason "WehhookServerFailed".
+- The k8s metric `webhook_admission_duration_seconds("<name of our new admission controller>", "validating", "put", "false")`
+    will show >1s latency and alert `WebhookAdmissionLatencyHigh` will fire.
+- Otherwise, we should be able to fall back on must gather.
+
+##### Disable the API extension (e.g. remove ValidatingWebhookConfiguration `<names of the our new webhook configs>`)
+
+- What consequences does it have on the cluster health?
+
+There are no consequences on cluster health for disabling the new ValidatingAdmissionWebhook(s).  They provide
+for better user experience with respect to error reporting when incorrect settings are provided Shared Resource instances
+or Pods (and their descendents) that specific CSI volumes using the Shared Resource CSI Driver.  But the controller / CSI Driver
+can also catch those errors, though as the Goals/User Stories note, the user has to do more digging.
+      
+
+- What consequences does it have on existing, running workloads?
+
+There are no consequences on existing, running workloads.  Only the Shared Resource Controller and CSI Driver impact
+existing, running workloads that have Shared Resource CSI Volume mounts.
+
+- What consequences does it have for newly created workloads?
+
+For properly constructed Shared Resources and Pods with Shared Resource CSI Volume Mounts, incorrect specifications
+will still be denied, but the errors will not manifest on Pod creation, but when Pod containers are trying to start.
+
+## Implementation History
+
+Initial submission of this EP
+
+## Drawbacks
+
+- Customers prove inexplicably resistant to reserving "openshift-" naming prefixes for Shared Resources (though they seem to have accepted similar restrictions for namespaces)
+
+## Alternatives
+
+First, currently see no reasons to use the older method (like in the early 3.0 days) of an admission plugin inside openshift-apiserver
+for validation, like you see for older API like Builds, DeploymentConfigs, ImageStreams, and Templates, could be found.
+But of course any SME guidance around this is welcome.
```

Comment:
I think a reason to use an admission plugin instead of webhook is the fact that we will be validating pods. Validating pods requires a level of availability that we already guarantee with openshift-apiserver. Openshift-apiserver already has a mutating admission webhook for imagestreams, this would not be that much different.

When I was working on the draft KEP for CSI volume security, we went down the path of adding an admission plugin to k8s precisely because we were validating pods. We can use feature gates to enable/disable the admission plugin.

Reply:
Right (forgot that precise element of the plugin/webhook comparison).  So a couple of follow ups @adambkaplan :
1) does it make sense to use validating webhooks for the SharedResource openshift-* name validation, or just use an admission plugin for both
2) I know we have examples of such admission plugins for the OpenShift API types in the OpenShift API server today, but I don't remember in the KEP CSI volume discussions, not that I was involved in all of them, if there is an example of OpenShift injecting one for a K8 type ... i.e. injecting one in the k8s API server.  Is that maybe how we prevent namespace creation from the command line with names starting with "openshift-" ?  Do you recall @adambkaplan 

I did not bring up this question with @deads2k during my exchange with him a few days ago.